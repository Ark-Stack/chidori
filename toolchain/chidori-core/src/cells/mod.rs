pub mod template_cell;
pub mod code_cell;
pub mod web_cell;
pub mod llm_prompt_cell;
pub mod memory_cell;
pub mod embedding_cell;
pub mod code_gen_cell;

use std::cmp::Ordering;
use std::collections::HashMap;
use ts_rs::TS;
use rkyv::{Archive, Deserialize, Serialize};
use serde_json::Value;
use crate::library::std::ai::llm::ChatModelBatch;

#[derive(
    TS,
    Archive,
    serde::Serialize,
    serde::Deserialize,
    Serialize,
    Deserialize,
    Debug,
    PartialEq,
    Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub enum SupportedLanguage {
    PyO3,
    Starlark,
    Deno,
}

#[derive(
    TS,
    Archive,
    serde::Serialize,
    serde::Deserialize,
    Serialize,
    Deserialize,
    Debug,
    PartialEq,
    Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub(crate) struct CodeCell {
    pub(crate) name: Option<String>,
    pub(crate) language: SupportedLanguage,
    pub(crate) source_code: String,
    pub(crate) function_invocation: Option<String>,
}


#[derive(
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub enum SupportedMemoryProviders {
    InMemory,
}


#[derive(
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub(crate) struct MemoryCell {
    pub(crate) name: Option<String>,
    pub(crate) provider: SupportedMemoryProviders,
    pub(crate) embedding_function: String,
}


#[derive(
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub(crate) struct TemplateCell {
    pub(crate) name: Option<String>,
    pub(crate) body: String,
}

#[derive(
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub(crate) struct WebserviceCellEndpoint {
    pub(crate) method: String,
    pub(crate) route: String,
    pub(crate) depended_function_identity: String,
    pub(crate) arg_mapping: Vec<(String, String)>,
}

#[derive(
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub(crate) struct WebserviceCell {
    pub(crate) name: Option<String>,
    pub(crate) configuration: String,
    pub(crate) port: u16,
}


#[derive(
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub(crate) struct ScheduleCell {
    pub(crate) configuration: String,
}


#[derive(
    TS,
    Archive,
    serde::Serialize,
    serde::Deserialize,
    Serialize,
    Deserialize,
    Debug,
    PartialEq,
    Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub enum SupportedModelProviders {
    OpenAI,
}


#[derive(
Default,
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub struct EjectionConfig {
    pub language: String,
    pub mode: String
}




#[derive(
Default,
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub struct LLMPromptCellChatConfiguration {
    pub(crate) import: Option<Vec<String>>,

    #[serde(rename = "fn")]
    pub(crate) function_name: Option<String>,

    pub model: String,
    pub frequency_penalty: Option<f64>,
    pub max_tokens: Option<i64>,
    pub presence_penalty: Option<f64>,
    pub stop: Option<Vec<String>>,
    pub temperature: Option<f64>,
    pub logit_bias: Option<HashMap<String, i32>>,
    pub user: Option<String>,
    pub seed: Option<i64>,
    pub top_p: Option<f64>,
}

#[derive(
    TS,
    Archive,
    serde::Serialize,
    serde::Deserialize,
    Serialize,
    Deserialize,
    Debug,
    PartialEq,
    Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub enum LLMPromptCell {
    Chat {
        function_invocation: bool,
        configuration: LLMPromptCellChatConfiguration,
        name: Option<String>,
        provider: SupportedModelProviders,
        req: String,
    },
    Completion {
        req: String,
    },
}


#[derive(
Default,
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub struct LLMCodeGenCellChatConfiguration {
    #[serde(rename = "fn")]
    pub(crate) function_name: Option<String>,

    pub model: String,
    pub frequency_penalty: Option<f64>,
    pub max_tokens: Option<i64>,
    pub presence_penalty: Option<f64>,
    pub stop: Option<Vec<String>>,
    pub temperature: Option<f64>,
    pub logit_bias: Option<HashMap<String, i32>>,
    pub user: Option<String>,
    pub seed: Option<i64>,
    pub top_p: Option<f64>,

    pub language: Option<String>,
}

#[derive(
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub struct LLMCodeGenCell {
    pub function_invocation: bool,
    pub configuration: LLMCodeGenCellChatConfiguration,
    pub name: Option<String>,
    pub provider: SupportedModelProviders,
    pub req: String,
}


#[derive(
TS,
Archive,
serde::Serialize,
serde::Deserialize,
Serialize,
Deserialize,
Debug,
PartialEq,
Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub struct LLMEmbeddingCell {
    pub function_invocation: bool,
    pub configuration: HashMap<String, String>,
    pub name: Option<String>,
    pub req: String,
}

#[derive(
    TS,
    Archive,
    serde::Serialize,
    serde::Deserialize,
    Serialize,
    Deserialize,
    Debug,
    PartialEq,
    Clone,
)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
#[ts(export, export_to = "package_node/types/")]
pub enum CellTypes {
    Code(CodeCell),
    CodeGen(LLMCodeGenCell),
    Prompt(LLMPromptCell),
    Embedding(LLMEmbeddingCell),
    Web(WebserviceCell),
    Template(TemplateCell),
    Memory(MemoryCell),
}

impl Eq for CellTypes {

}

impl PartialOrd for CellTypes {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(Ordering::Equal)
    }
}

impl Ord for CellTypes {
    fn cmp(&self, other: &Self) -> Ordering {
        self.partial_cmp(other).unwrap_or(Ordering::Equal)
    }
}

pub fn get_cell_name(cell: &CellTypes) -> &Option<String> {
    match &cell {
        CellTypes::Code(c) => &c.name,
        CellTypes::Prompt(c) => match c {
            LLMPromptCell::Chat { name, .. } => name,
            LLMPromptCell::Completion { .. } => &None,
        },
        CellTypes::Web(c) => &c.name,
        CellTypes::Template(c) => &c.name,
        CellTypes::Memory(c) => &c.name,
        CellTypes::Embedding(c) => &c.name,
        CellTypes::CodeGen(c) => &c.name
    }
}
