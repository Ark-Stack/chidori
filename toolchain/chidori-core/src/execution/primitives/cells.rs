use rkyv::{
    archived_root, check_archived_root,
    ser::{serializers::AllocSerializer, Serializer},
    Archive, Deserialize, Serialize,
};
use serde_json::Value;
use std::collections::HashMap;

#[derive(Archive, Serialize, Deserialize, Debug, PartialEq, Clone)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
pub enum SupportedLanguage {
    Python,
    Starlark,
    Deno,
}

#[derive(Archive, Serialize, Deserialize, Debug, PartialEq, Clone)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
pub(crate) struct CodeCell {
    pub(crate) language: SupportedLanguage,
    pub(crate) source_code: String,
    pub(crate) function_invocation: Option<String>,
}

#[derive(Archive, Serialize, Deserialize, Debug, PartialEq, Clone)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
pub enum SupportedModelProviders {
    OpenAI,
}

#[derive(Archive, Serialize, Deserialize, Debug, PartialEq, Clone)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
pub enum LLMPromptCell {
    Chat {
        path: Option<String>,
        provider: SupportedModelProviders,
        req: String,
    },
    Completion {
        req: String,
    },
    Embedding {
        req: String,
    },
}

#[derive(Archive, Serialize, Deserialize, Debug, PartialEq, Clone)]
#[archive(bound(serialize = "__S: rkyv::ser::ScratchSpace + rkyv::ser::Serializer"))]
#[archive(check_bytes)]
#[archive_attr(check_bytes(
    bound = "__C: rkyv::validation::ArchiveContext, <__C as rkyv::Fallible>::Error: std::error::Error"
))]
#[archive_attr(derive(Debug))]
pub enum CellTypes {
    Code(CodeCell),
    Prompt(LLMPromptCell),
}
