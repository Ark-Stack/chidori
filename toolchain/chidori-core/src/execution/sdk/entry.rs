/// This is an SDK for building execution graphs. It is designed to be used iteratively.

/// Add a block of code to a runtime execution graph
fn load_module() {}

/// Execute a target directory of code. Specify your entrypoint into the code base.
fn parse_folder() {}
