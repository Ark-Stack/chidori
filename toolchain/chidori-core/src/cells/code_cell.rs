use futures_util::FutureExt;
use crate::cells::{CodeCell, SupportedLanguage};
use crate::execution::execution::ExecutionState;
use crate::execution::primitives::operation::{InputItemConfiguation, InputSignature, InputType, OperationNode, OutputItemConfiguation, OutputSignature};

/// Code cells allow notebooks to evaluate source code in a variety of languages.
#[tracing::instrument]
pub fn code_cell(cell: &CodeCell) -> OperationNode {
    match cell.language {
        SupportedLanguage::PyO3 => {
            // TODO: all callable functions should be exposed as their own OperationNodes
            //       these can be depended on throughout the graph as their own cells.
            let paths =
                chidori_static_analysis::language::python::parse::extract_dependencies_python(
                    &cell.source_code,
                );
            let report = chidori_static_analysis::language::python::parse::build_report(&paths);

            let mut input_signature = InputSignature::new();
            for (key, value) in &report.cell_depended_values {
                input_signature.globals.insert(
                    key.clone(),
                    InputItemConfiguation {
                        ty: Some(InputType::String),
                        default: None,
                    },
                );
            }

            let mut output_signature = OutputSignature::new();
            for (key, value) in &report.cell_exposed_values {
                output_signature.globals.insert(
                    key.clone(),
                    OutputItemConfiguation {
                        ty: Some(InputType::String),
                    },
                );
            }

            for (key, value) in &report.triggerable_functions {
                output_signature.functions.insert(
                    key.clone(),
                    OutputItemConfiguation {
                        ty: Some(InputType::Function),
                    },
                );
            }

            let cell = cell.clone();
            OperationNode::new(
                cell.name.clone(),
                input_signature,
                output_signature,
                Box::new(move |s, x, _, _| {
                    let cell = cell.clone();
                    // TODO: this needs to handle stdout and errors
                    let s = s.clone();
                    async move {
                        crate::library::std::code::runtime_pyo3::source_code_run_python(
                            &s,
                            &cell.source_code,
                            &x,
                            &cell.function_invocation,
                        ).await
                            .unwrap()
                            .0
                    }.boxed()
                }),
            )
        }
        SupportedLanguage::Starlark => OperationNode::new(
            cell.name.clone(),
            InputSignature::new(),
            OutputSignature::new(),
            Box::new(|_, x, _, _| async move { x}.boxed()),
        ),
        SupportedLanguage::Deno => {
            let paths =
                chidori_static_analysis::language::javascript::parse::extract_dependencies_js(
                    &cell.source_code,
                );
            let report = chidori_static_analysis::language::javascript::parse::build_report(&paths);

            let mut input_signature = InputSignature::new();
            for (key, value) in &report.cell_depended_values {
                input_signature.globals.insert(
                    key.clone(),
                    InputItemConfiguation {
                        ty: Some(InputType::String),
                        default: None,
                    },
                );
            }

            let mut output_signature = OutputSignature::new();
            for (key, value) in &report.cell_exposed_values {
                output_signature.globals.insert(
                    key.clone(),
                    OutputItemConfiguation {
                        ty: Some(InputType::String),
                    },
                );
            }

            for (key, value) in &report.triggerable_functions {
                output_signature.functions.insert(
                    key.clone(),
                    OutputItemConfiguation {
                        ty: Some(InputType::Function),
                    },
                );
            }

            let cell = cell.clone();
            OperationNode::new(
                cell.name.clone(),
                input_signature,
                output_signature,
                Box::new(move |_, x, _, _| {
                    // TODO: this needs to handle stdout and errors
                    let cell = cell.clone();
                    async move {
                        let result = tokio::task::spawn_blocking(move || {
                            let runtime = tokio::runtime::Runtime::new().unwrap();
                            let result = runtime.block_on(crate::library::std::code::runtime_deno::source_code_run_deno(
                                &ExecutionState::new(),
                                &cell.source_code,
                                &x,
                                &cell.function_invocation,
                            ));
                            match result {
                                Ok(v) => v.0,
                                Err(e) => panic!("{:?}", e),
                            }
                        }).await.unwrap();
                        result
                    }.boxed()
                }),
            )
        }
    }
}


#[cfg(test)]
mod test {
    #[tokio::test]
    async fn test_code_cell() {


    }
}