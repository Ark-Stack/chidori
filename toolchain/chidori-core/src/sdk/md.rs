use crate::sdk::entry::Chidori;
use chidori_prompt_format::extract_yaml_frontmatter_string;
use indoc::indoc;
use std::collections::HashMap;
use std::path::Path;
use serde_derive::Serialize;
use thiserror::Error;
use crate::cells::{CellTypes, CodeCell, LLMCodeGenCell, LLMEmbeddingCell, LLMPromptCell, MemoryCell, SupportedLanguage, SupportedMemoryProviders, SupportedModelProviders, TemplateCell, TextRange, WebserviceCell};

#[derive(PartialEq, Serialize, Debug)]
pub struct MarkdownCodeBlock {
    pub tag: String,
    pub name: Option<String>,
    pub body: String,
    pub range: TextRange,
}

enum CodeResource {
    Python,
    Js,
    Markdown
}

#[derive(Debug)]
pub struct ParsedFile {
    filename: Option<Box<std::path::PathBuf>>,
    code: Option<String>,
    num_lines: usize,
    pub(crate) result: Vec<MarkdownCodeBlock>,
}

pub(crate) fn extract_code_blocks(body: &str) -> Vec<MarkdownCodeBlock> {
    let mut code_blocks = Vec::new();
    let mut start = 0;

    // Iterate over each occurrence of backticks
    while let Some(end) = body[start..].find("```") {
        start += end + 3; // Move start to the character after the closing ```

        if let Some(end_of_code) = body[start..].find("```") {
            let code = &body[start..start + end_of_code].trim();

            // Extract first line to separate tag and name
            let mut lines = code.lines();
            let first_line = lines.next().unwrap_or_default();
            let rest: String = lines.collect::<Vec<&str>>().join("\n");

            let tag_and_name: Vec<&str> = first_line.split_whitespace().collect();
            let tag = tag_and_name.get(0).cloned().unwrap_or_default().to_string();
            let name = tag_and_name.get(1).and_then(|n| n.strip_prefix('(').and_then(|n| n.strip_suffix(')'))).map(|n| n.to_string());

            // Add the code block with the text range
            code_blocks.push(MarkdownCodeBlock {
                tag,
                name,
                body: rest,
                range: TextRange {
                    start,
                    end: start + end_of_code
                },
            });

            start += end_of_code + 3; // Move start to the character after the closing ```
        } else {
            break; // No closing backticks found, exit the loop
        }
    }

    code_blocks
}


fn parse_markdown_file(filename: &Path) -> ParsedFile {
    match std::fs::read_to_string(filename) {
        Err(e) => ParsedFile {
            filename: Some(Box::new(filename.to_path_buf())),
            code: Some("".to_owned()),
            num_lines: 0,
            result: vec![],
        },
        Ok(source) => {
            let num_lines = source.lines().count();
            let result = extract_code_blocks(&source);
            ParsedFile {
                filename: Some(Box::new(filename.to_path_buf())),
                code: Some(source.to_string()),
                num_lines,
                result,
            }
        }
    }
}

pub fn load_folder(path: &Path) -> anyhow::Result<Vec<ParsedFile>> {
    let mut res = vec![];
    for entry in path.read_dir()? {
        let entry = entry?;
        let metadata = entry.metadata()?;

        let path = entry.path();
        if metadata.is_dir() {
            res.extend(load_folder(&path)?);
        }

        if metadata.is_file() && path.extension().and_then(|s| s.to_str()) == Some("md") {
            let parsed_file = parse_markdown_file(&path);
            res.push(parsed_file);
        }

        if metadata.is_file() && path.extension().and_then(|s| s.to_str()) == Some("py") {
            let parsed_file = parse_markdown_file(&path);
            res.push(parsed_file);
        }

        if metadata.is_file() && path.extension().and_then(|s| s.to_str()) == Some("js") {
            let parsed_file = parse_markdown_file(&path);
            res.push(parsed_file);
        }

        if metadata.is_file() && path.extension().and_then(|s| s.to_str()) == Some("ts") {
            let parsed_file = parse_markdown_file(&path);
            res.push(parsed_file);
        }
    }
    Ok(res)
}

#[derive(Error, Debug)]
pub enum InterpretError {
    #[error("Failed to split frontmatter: {0}")]
    FrontmatterSplitError(String),
    #[error("Failed to deserialize YAML: {0}")]
    YamlDeserializeError(#[from] serde_yaml::Error),
    #[error("Failed to parse port number")]
    PortParseError,
}


pub fn interpret_code_block(block: &MarkdownCodeBlock) -> Result<Option<CellTypes>, InterpretError> {
    let (frontmatter, body) = chidori_prompt_format::templating::templates::split_frontmatter(&block.body)
        .map_err(|e| InterpretError::FrontmatterSplitError(e.to_string()))?;

    Ok(match block.tag.as_str() {
        "python" | "javascript" | "py" | "js" | "ts" | "typescript" => {
            let language = match block.tag.as_str() {
                "python" | "py" => SupportedLanguage::PyO3,
                "javascript" | "js" | "typescript" | "ts" => SupportedLanguage::Deno,
                _ => unreachable!(), // Given the outer match, this branch should never be reached
            };
            Some(CellTypes::Code(CodeCell {
                name: block.name.clone(),
                language,
                source_code: block.body.clone(),
                function_invocation: None,
            }, block.range.clone()))
        },
        "memory" => Some(CellTypes::Memory(MemoryCell {
            name: block.name.clone(),
            provider: SupportedMemoryProviders::InMemory,
            embedding_function: block.body.clone(),
        }, block.range.clone())),
        "embedding" => Some(CellTypes::Embedding(LLMEmbeddingCell {
            function_invocation: false,
            configuration: serde_yaml::from_str(&frontmatter)?,
            name: block.name.clone(),
            req: body,
        }, block.range.clone())),
        "prompt" => Some(CellTypes::Prompt(LLMPromptCell::Chat {
            function_invocation: false,
            configuration: serde_yaml::from_str(&frontmatter)?,
            name: block.name.clone(),
            provider: SupportedModelProviders::OpenAI,
            req: body,
        }, block.range.clone())),
        "codegen" => Some(CellTypes::CodeGen(LLMCodeGenCell {
            function_invocation: false,
            configuration: serde_yaml::from_str(&frontmatter)?,
            name: block.name.clone(),
            provider: SupportedModelProviders::OpenAI,
            req: body,
        }, block.range.clone())),
        "html" => Some(CellTypes::Template(TemplateCell {
            name: block.name.clone(),
            body: block.body.clone(),
        }, block.range.clone())),
        _ => None,
    })
}


#[cfg(test)]
mod test {
    use super::*;
    use crate::execution::primitives::serialized_value::RkyvObjectBuilder;
    use indoc::indoc;
    use std::collections::HashMap;
    use std::fs;

    #[test]
    fn test_core1() {
        let contents = fs::read_to_string("./examples/core1_simple_math/core.md")
            .expect("Should have been able to read the file");
        let v: Vec<Option<CellTypes>> = extract_code_blocks(&contents)
            .iter()
            .flat_map(|block| interpret_code_block(block).ok())
            .collect();
        insta::with_settings!({
        }, {
            insta::assert_yaml_snapshot!(v);
        });
    }

    #[test]
    fn test_core2() {
        let contents = fs::read_to_string("./examples/core2_marshalling/core.md")
            .expect("Should have been able to read the file");
        let v: Vec<Option<CellTypes>> = extract_code_blocks(&contents)
            .iter()
            .flat_map(|block| interpret_code_block(block).ok())
            .collect();
        insta::with_settings!({
        }, {
            insta::assert_yaml_snapshot!(v);
        });
    }

    #[test]
    fn test_core3() {
        let contents = fs::read_to_string("./examples/core3_function_invocations/core.md")
            .expect("Should have been able to read the file");
        let v: Vec<Option<CellTypes>> = extract_code_blocks(&contents)
            .iter()
            .flat_map(|block| interpret_code_block(block).ok())
            .collect();
        insta::with_settings!({
        }, {
            insta::assert_yaml_snapshot!(v);
        });
    }

    #[test]
    fn test_core4() {
        let contents = fs::read_to_string("./examples/core4_async_function_invocations/core.md")
            .expect("Should have been able to read the file");
        let v: Vec<Option<CellTypes>> = extract_code_blocks(&contents)
            .iter()
            .flat_map(|block| interpret_code_block(block).ok())
            .collect();
        insta::with_settings!({
        }, {
            insta::assert_yaml_snapshot!(v);
        });
    }

    #[test]
    fn test_core5() {
        let contents = fs::read_to_string("./examples/core5_prompts_invoked_as_functions/core.md")
            .expect("Should have been able to read the file");
        let v: Vec<Option<CellTypes>> = extract_code_blocks(&contents)
            .iter()
            .flat_map(|block| interpret_code_block(block).ok())
            .collect();
        insta::with_settings!({
        }, {
            insta::assert_yaml_snapshot!(v);
        });
    }

    #[test]
    fn test_core6() {
        let contents = fs::read_to_string("./examples/core6_prompts_leveraging_function_calling/core.md")
            .expect("Should have been able to read the file");
        let v: Vec<Option<CellTypes>> = extract_code_blocks(&contents)
            .iter()
            .flat_map(|block| interpret_code_block(block).ok())
            .collect();
        insta::with_settings!({
        }, {
            insta::assert_yaml_snapshot!(v);
        });
    }

    #[test]
    fn test_core7() {
        let contents = fs::read_to_string("./examples/core7_rag_stateful_memory_cells/core.md")
            .expect("Should have been able to read the file");
        let v: Vec<Option<CellTypes>> = extract_code_blocks(&contents)
            .iter()
            .flat_map(|block| interpret_code_block(block).ok())
            .collect();
        insta::with_settings!({
        }, {
            insta::assert_yaml_snapshot!(v);
        });
    }

    #[test]
    fn test_extract_markdown() {
        let extracted = extract_code_blocks(indoc! {  r#"
        Generation
        
        ```python
        y = 20
        def add(a, b):
            return a + b
        ```

        ```javascript
        ---
        a: 2
        ---
        const x = add(2,2);
        ```
        
        ```prompt (multi_prompt)
        Multiply {y} times {x}
        ```

        ```html (named_html)
        <div>Example</div>
        ```
        "#
        });

        let mut map = HashMap::new();
        map.insert("a".to_string(), "2".to_string());
        insta::with_settings!({
        }, {
            insta::assert_yaml_snapshot!(extracted);
        });
    }
}
