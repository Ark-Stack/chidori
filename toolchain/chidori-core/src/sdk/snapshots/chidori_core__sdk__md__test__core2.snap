---
source: chidori-core/src/sdk/md.rs
expression: v
---
- Code:
    - name: ~
      language: PyO3
      source_code: "x0 = 1\nx1 = \"string\"\nx2 = [1, 2, 3]\nx3 = {\"a\": 1, \"b\": 2, \"c\": 3}\nx4 = False\nx5 = 1.0\nx6 = (1, 2, 3)\nx7 = {\"a\", \"b\", \"c\"}\n\n# def example(x):\n#     return x\n\n# TODO: marshalling of functions is not currently supported\n# x8 = example\n\n# TODO: marshalling of classes is not currently supported\n# class ClassExample:\n#     def __init__(self, x):\n#         self.x = x\n#         \n# x9 = ClassExample"
      function_invocation: ~
    - start: 162
      end: 564
- Code:
    - name: ~
      language: Deno
      source_code: "Chidori.assertEq(x0, 1);\nChidori.assertEq(x1, \"string\");\nChidori.assertEq(x2, [1,2,3]);\nChidori.assertEq(x3, {\"a\": 1, \"b\": 2, \"c\": 3});\nChidori.assertEq(x4, false);\nChidori.assertEq(x5, 1.0);\nChidori.assertEq(x6, [1, 2, 3]);\n\n// TODO: marshalling of sets is not currently supported\n// Chidori.assertEq(x7, [\"c\", \"b\", \"a\"]);\n\nChidori.assertEq(x8, null);\n\n// TODO: marshalling of functions is not currently supported\n// Chidori.assertEq(typeof x9, \"function\");"
      function_invocation: ~
    - start: 572
      end: 1042
- Code:
    - name: ~
      language: Deno
      source_code: "const y0 = 1;\nconst y1 = \"string\";\nconst y2 = [1, 2, 3];\nconst y3 = {a: 1, b: 2, c: 3};\nconst y4 = false;\nconst y5 = 1.0;\nconst y6 = [1, 2, 3];\n// TODO: marshalling of sets is not currently supported\n// const y7 = new Set([\"a\", \"b\", \"c\"]);\n\n// TODO: marshalling of functions is not currently supported\n// const y8 = (y) => y;\n\n// TODO: marshalling of classes is not currently supported\n// class ClassExample {\n//    constructor(y) {\n//        this.y = y;\n//    }\n//};\n//const y9 = ClassExample;"
      function_invocation: ~
    - start: 1095
      end: 1601
- Code:
    - name: ~
      language: PyO3
      source_code: "import unittest\n\nclass TestMarshalledValues(unittest.TestCase):\n    def test_all(self):\n        self.assertEqual(y0, 1)\n        self.assertEqual(y1, \"string\")\n        self.assertEqual(y2, [1,2,3])\n        self.assertEqual(y3, {\"a\": 1, \"b\": 2, \"c\": 3})\n        self.assertEqual(y4, False)\n        self.assertEqual(y5, 1.0)\n        self.assertEqual(y6, [1,2,3])\n\nunittest.TextTestRunner().run(unittest.TestLoader().loadTestsFromTestCase(TestMarshalledValues))"
      function_invocation: ~
    - start: 1609
      end: 2074
